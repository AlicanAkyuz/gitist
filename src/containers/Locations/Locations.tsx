import React, { FunctionComponent } from 'react';
import { Query } from 'react-apollo';
import { GET_LOCATIONS } from '../../queries';

import Location from '../../components/Location';
import Loading from '../../components/Loading';
import Error from '../../components/Error';

interface Data {
  locations: [
    {
      name: string;
      totalDevelopers: number;
      totalRepositories: number;
      languageUsage: [
        {
          language: {
            name: string;
          };
        },
      ];
    },
  ];
}

interface Variables {
  limit: number;
  offset: number;
  orderBy: object;
}

const Locations: React.FunctionComponent = () => (
  <Query<Data, Variables>
    query={GET_LOCATIONS}
    variables={{
      limit: 5,
      offset: 0,
      orderBy: { direction: 'DESC', field: 'SCORE' },
    }}
  >
    {({ data, loading, error, fetchMore }) => {
      if (error) return <Error />;
      if (loading || !data) return <Loading />;
      let offset = 0;
      const { locations } = data;
      return (
        <Location
          locations={locations}
          onLoadMore={() => {
            offset = offset + 5;
            return fetchMore({
              variables: { offset },
              updateQuery: (prev, { fetchMoreResult }) => {
                if (!fetchMoreResult) return prev;
                let newData = JSON.parse(JSON.stringify(prev));
                newData.locations = Object.assign(
                  [],
                  [...prev.locations, ...fetchMoreResult.locations],
                );
                return newData;
              },
            });
          }}
        />
      );
    }}
  </Query>
);
export default Locations;
