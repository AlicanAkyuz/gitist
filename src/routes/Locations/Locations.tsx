import React, { FunctionComponent } from 'react';
import { Query } from 'react-apollo';
import { GET_LOCATIONS } from '../../queries';

import LocationList from '../../components/LocationList';
import Loading from '../../components/Loading';
import Error from '../../components/Error';

interface Data {
  locations: [
    {
      name: string;
      totalDevelopers: number;
      totalRepositories: number;
      languageUsage: [
        {
          language: {
            name: string;
          };
        },
      ];
    },
  ];
}

interface Variables {
  limit: number;
  offset: number;
  orderBy: object;
}

const Locations: FunctionComponent = () => (
  <Query<Data, Variables>
    query={GET_LOCATIONS}
    variables={{
      limit: 5,
      offset: 0,
      orderBy: { direction: 'DESC', field: 'SCORE' },
    }}
  >
    {({ data, loading, error, fetchMore }) => {
      if (error) return <Error />;
      if (loading || !data) return <Loading />;
      let offset = 0;
      const { locations } = data;
      return (
        <LocationList
          locations={locations}
          onLoadMore={async () => {
            try {
              offset = offset + 5;
              await fetchMore({
                variables: { offset },
                updateQuery: (prev, { fetchMoreResult }) => {
                  if (!fetchMoreResult) return prev;
                  let newData = JSON.parse(JSON.stringify(prev));
                  newData.locations = Object.assign(
                    [],
                    [...prev.locations, ...fetchMoreResult.locations],
                  );
                  return newData;
                },
              });
            } catch (err) {
              console.log('err');
              // When the variables of <Query> change while a fetchMore is still in progress, we get this error: ObservableQuery with this id does not exist: id on unmounted component, which is currently an open issue at: https://github.com/apollographql/apollo-client/issues/4114
            }
          }}
        />
      );
    }}
  </Query>
);
export default Locations;
