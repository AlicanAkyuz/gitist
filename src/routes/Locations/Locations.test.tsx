import React from 'react';
import { shallow } from 'enzyme';

import Locations from './Locations';

it('Locations corrently renders a navbar and Routes component', () => {
  const wrapper = shallow(<Locations />);
  expect(wrapper).toMatchSnapshot();
});
