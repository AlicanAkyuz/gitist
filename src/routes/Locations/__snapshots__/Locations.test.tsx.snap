// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locations corrently renders a navbar and Routes component 1`] = `
<Query
  query={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "locations",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "orderBy",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "orderBy",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "limit",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "limit",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "offset",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "offset",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "locations",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "totalDevelopers",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "totalRepositories",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "limit",
                          },
                          "value": Object {
                            "kind": "IntValue",
                            "value": "1",
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "languageUsage",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "language",
                            },
                            "selectionSet": Object {
                              "kind": "SelectionSet",
                              "selections": Array [
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "name",
                                  },
                                  "selectionSet": undefined,
                                },
                              ],
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "LocationOrder",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "orderBy",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "Int",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "limit",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "Int",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "offset",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 301,
        "start": 0,
      },
    }
  }
  variables={
    Object {
      "limit": 5,
      "offset": 0,
      "orderBy": Object {
        "direction": "DESC",
        "field": "SCORE",
      },
    }
  }
>
  <Component />
</Query>
`;
